samcli/commands/package/artifact_exporter.py

```diff
@@ -0,0 +1,611 @@
+    # up the file before uploading This is useful for Lambda functions.
+    FORCE_ZIP = False
+
+    def __init__(self, uploader):
+        self.uploader = uploader
+
+    def export(self, resource_id, resource_dict, parent_dir):
+        if resource_dict is None:
+            return
+
+        property_value = jmespath.search(self.PROPERTY_NAME, resource_dict)
+
+        if not property_value and not self.PACKAGE_NULL_PROPERTY:
+            return
+
+        if isinstance(property_value, dict):
+            LOG.debug("Property %s of %s resource is not a URL", self.PROPERTY_NAME, resource_id)
+            return
+
+        # If property is a file but not a zip file, place file in temp
+        # folder and send the temp folder to be zipped
+        temp_dir = None
+        if is_local_file(property_value) and not is_zip_file(property_value) and self.FORCE_ZIP:
+            temp_dir = copy_to_temp_dir(property_value)
+            set_value_from_jmespath(resource_dict, self.PROPERTY_NAME, temp_dir)
+
+        try:
+            self.do_export(resource_id, resource_dict, parent_dir)
+
+        except Exception as ex:
+            LOG.debug("Unable to export", exc_info=ex)
+            raise exceptions.ExportFailedError(
+                resource_id=resource_id, property_name=self.PROPERTY_NAME, property_value=property_value, ex=ex
+            )
+        finally:
+            if temp_dir:
+                shutil.rmtree(temp_dir)
+
+    def do_export(self, resource_id, resource_dict, parent_dir):
+        """
+        Default export action is to upload artifacts and set the property to
+        S3 URL of the uploaded object
+        """
+        uploaded_url = upload_local_artifacts(resource_id, resource_dict, self.PROPERTY_NAME, parent_dir, self.uploader)
+        set_value_from_jmespath(resource_dict, self.PROPERTY_NAME, uploaded_url)
+
+
+class ResourceWithS3UrlDict(Resource):
+    """
+    Represents CloudFormation resources that need the S3 URL to be specified as
+    an dict like {Bucket: "", Key: "", Version: ""}
+    """
+
+    BUCKET_NAME_PROPERTY = None
+    OBJECT_KEY_PROPERTY = None
+    VERSION_PROPERTY = None
+
+    def do_export(self, resource_id, resource_dict, parent_dir):
+        """
+        Upload to S3 and set property to an dict representing the S3 url
+        of the uploaded object
+        """
+
+        artifact_s3_url = upload_local_artifacts(
+            resource_id, resource_dict, self.PROPERTY_NAME, parent_dir, self.uploader
+        )
+
+        parsed_url = parse_s3_url(
+            artifact_s3_url,
+            bucket_name_property=self.BUCKET_NAME_PROPERTY,
+            object_key_property=self.OBJECT_KEY_PROPERTY,
+            version_property=self.VERSION_PROPERTY,
+        )
+        set_value_from_jmespath(resource_dict, self.PROPERTY_NAME, parsed_url)
+
+
+class ServerlessFunctionResource(Resource):
+    RESOURCE_TYPE = "AWS::Serverless::Function"
```

Comment:
In the help text of the command, should we be adding what resources are supported?

How do we keep this in sync with https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/commands/_utils/template.py#L19

Reply:
Can do that.